connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+        "_id": "Id",
+        "_timestamp": "LastItemModifiedDate",
+        "creation_time": "Created",
+    },
+    LIST_ITEM: {
+        "title": "Title",
+        "author_id": "EditorId",
+        "creation_time": "Created",
+        "_timestamp": "Modified",
+    },
+    DRIVE_ITEM: {
+        "title": "Name",
+        "creation_time": "TimeCreated",
+        "_timestamp": "TimeLastModified",
+    },
+}
+
+
+def encode(original_string):
+    """Performs encoding on the name of objects
+    containing special characters in their url, and
+    replaces single quote with two single quote since quote
+    is treated as an escape character
+
+    Args:
+        original_string(string): String containing special characters
+
+    Returns:
+        encoded_string(string): Parsed string without single quotes
+    """
+    encoded_string = urllib.parse.quote(original_string, safe="'")
+    return encoded_string.replace("'", "''")
+
+
+def get_expires_at(expires_in):
+    """Adds seconds to the current utc time.
+
+    Args:
+        expires_in (int): Seconds after which the token will expire.
+    """
+    expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+    # account for clock skew
+    expires_at -= timedelta(seconds=20)
+    return iso_utc(when=expires_at)
+
+
+def is_expired(expires_at):
+    """Compares the given time with present time
+
+    Args:
+        expires_at (datetime): Time of token expires.
+    """
+    # Recreate in case there's no expires_at present
+    if expires_at is None:
+        return True
+    if not isinstance(expires_at, datetime):
+        expires_at = datetime.fromisoformat(expires_at)
+    if datetime.utcnow() >= expires_at:
+        return True
+    else:
+        return False
+
+
+class SharepointDataSource(BaseDataSource):
+    """Sharepoint"""
+
+    name = "Sharepoint"
+    service_type = "sharepoint"
+
+    def __init__(self, configuration):
+        """Setup the connection to the Sharepoint
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self.is_cloud = self.configuration["is_cloud"]
+        self.site_collections = self.configuration["site_collections"]
+        self.ssl_disabled = self.configuration["ssl_disabled"]
+        self.host_url = self.configuration["host_url"]
+        self.certificate = self.configuration["ssl_ca"]
+        self.enable_content_extraction = self.configuration["enable_content_extraction"]
+        self.retry_count = self.configuration["retry_count"]
+        self.ssl_ctx = False
+        self.session = None
+        self.access_token = None
+        self.token_expires_in = None
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for Sharepoint
+
+        Returns:
+            dictionary: Default configuration.
+        """
+        return {
```

Comment:
What's our stance on default field values @tarekziade @danajuratoni?

We have them here for FTests, but after @timgrein does his changes we can remove them or provide defaults only for the fields that we need.

Reply:
why is it a problem to have a default value for each field? it allows the connector to work out of the box for the e2e test

Reply:
It's displayed in UI and most of the time it does not help - users will just delete them. Some of these values make sense as reasonable defaults, but some don't IMO - such as `password`, `client_id`, `client_secret`, `tenant`, `site_collections`. We also set `ssl_disabled=True` for tests but for real users we probably should default it to False instead.

Reply:
Defaults should not be required for fields. I agree it would be good to set reasonable defaults especially for fields such as `retry_count`, but requiring users to delete `password` before being able to enter their own, is not a good user experience.

Reply:
@danajuratoni did I get it right that we need to ask @akanshi-crest to remove some of the default fields' values?

Reply:
Yes, let's remove defaults that are not useful. We'll get a better user experience via rich configurable fields coming in 8.8 where we will handle adding placeholders, tooltips, etc.

Reply:
By "useful defaults" I mean those for fields used to fine tune a connector such as `retry_count` or `fetch_size` or disabled SSL for example.

Reply:
I have removed the non useful defaults from configurable fields such as `client_id`, `secret_id`, `tenant` and `tenant_id` but `username` and `password` values are required and we can't remove it as it is used to run e2e tests.